### ISSUE TITLE
[css-grid] grid-gap is not retained through media queries

### ISSUE THREAD

**Issue by Dan503** [postcss/autoprefixer#1045]
I wrote this test css-grid code and realised that the `grid-gap` functionality wasn't getting carried through media queries.

````css
/* changing a grid-template with a media-query */
.grid {
  display: grid;
  grid-gap: 10px; /* grid-gap defined here */
  grid-template:
    "a   b" 100px
    "c   d" 100px
    "e   f" 100px /
    1fr  1fr;
}

.cell-A {
  grid-area: a;
}

.cell-B {
  grid-area: b;
}

.cell-C {
  grid-area: c;
}

.cell-D {
  grid-area: d;
}

.cell-E {
  grid-area: e;
}

.cell-F {
  grid-area: f;
}

@media (min-width: 600px){
  .grid {
    /* grid-gap setting should be remembered here */
    grid-template-areas:
      "a   b   c"
      "d   e   f";
    grid-template-columns: repeat(3, 1fr);
    grid-template-rows: repeat(2, 100px);
  }
}
````
It compiles into this:
````css
.grid {
  display: -ms-grid;
  display: grid;
  grid-gap: 10px;
  -ms-grid-rows: 100px 10px 100px 10px 100px;
  -ms-grid-columns: 1fr 10px 1fr;
      grid-template: "a   b" 100px
 "c   d" 100px
 "e   f" 100px /
 1fr  1fr;
}

.cell-A {
  -ms-grid-row: 1;
  -ms-grid-column: 1;
  grid-area: a;
}

.cell-B {
  -ms-grid-row: 1;
  -ms-grid-column: 3;
  grid-area: b;
}

.cell-C {
  -ms-grid-row: 3;
  -ms-grid-column: 1;
  grid-area: c;
}

.cell-D {
  -ms-grid-row: 3;
  -ms-grid-column: 3;
  grid-area: d;
}

.cell-E {
  -ms-grid-row: 5;
  -ms-grid-column: 1;
  grid-area: e;
}

.cell-F {
  -ms-grid-row: 5;
  -ms-grid-column: 3;
  grid-area: f;
}

@media (min-width: 600px) {
  .grid {
        grid-template-areas: "a   b   c"
 "d   e   f";

    /* !!! IT HAS FORGOTTEN ABOUT THE GRID-GAP SETTING !!! */

    -ms-grid-columns: (1fr)[3];
    grid-template-columns: repeat(3, 1fr);
    -ms-grid-rows: (100px)[2];
    grid-template-rows: repeat(2, 100px);
  }
  .cell-A {
    -ms-grid-row: 1;
    -ms-grid-column: 1;
  }
  .cell-B {
    -ms-grid-row: 1;
    -ms-grid-column: 2;
  }
  .cell-C {
    -ms-grid-row: 1;
    -ms-grid-column: 3;
  }
  .cell-D {
    -ms-grid-row: 2;
    -ms-grid-column: 1;
  }
  .cell-E {
    -ms-grid-row: 2;
    -ms-grid-column: 2;
  }
  .cell-F {
    -ms-grid-row: 2;
    -ms-grid-column: 3;
  }
}
````

My idea is that if the `grid-area` name is consistent throughout various media-queries then it is safe to assume that it is looking at the same grid.

If the `grid-gap` setting is defined outside of a media query, then that gap setting needs to be applied as a default across all grids that share that same area name.

````css
.grid {
  display: grid;
  grid-gap: 10px;/* grid-gap defined */
  grid-template:
    "a   b" 100px
    "c   d" 100px
    "e   f" 100px /
    1fr  1fr;
}


@media (min-width: 600px){
  .grid {
    /* grid-gap: 10px; << grid-gap inherited */
    grid-template-areas:
      "a   b   c"
      "d   e   f";
    grid-template-columns: repeat(3, 1fr);
    grid-template-rows: repeat(2, 100px);
  }
}
````

If a `grid-gap` setting is applied inside of a media query, it will only be applied as a default to other grids that live inside media queries that agree with the `grid-gap` media query.

````css
.grid {
  display: grid;
  /* grid-gap ignored */
  grid-template:
    "a   b" 100px
    "c   d" 100px
    "e   f" 100px /
    1fr  1fr;
}


@media (min-width: 600px){
  .grid {
    grid-gap: 10px;/* grid-gap defined */
    grid-template-areas:
      "a   b   c"
      "d   e   f";
    grid-template-columns: repeat(3, 1fr);
    grid-template-rows: repeat(2, 100px);
  }
}

@media (min-width: 900px){
  .grid {
    /* grid-gap: 10px; << grid-gap inherited */
    grid-template-areas:
      "a   b   c   d   e   f";
    grid-template-columns: repeat(6, 1fr);
    grid-template-rows: 100px;
  }
}

@media (max-width: 400px){
  .grid {
    /* grid-gap ignored */
    grid-template-areas:
      "a"
      "b"
      "c"
      "d"
      "e"
      "f";
    grid-template-columns: 1fr;
    grid-template-rows: repeat(6, 100px);
  }
}
````

If there is a conflict then the `grid-gap` setting that is defined last in the style sheet should have the priority.

**Comment by yepninja** [comment]
> My idea is that if the grid-area name is consistent throughout various media-queries then it is safe to assume that it is looking at the same grid.
To my mind, it's not safe. People can use the same names for different grids.
Maybe it's better to create a separate postcss plugin to handle this case.
@ai What do you think?

**Comment by ai** [comment]
We need at least warm user and recommend the plugin in this warning

**Comment by Dan503** [comment]
> To my mind, it's not safe

Isn't the whole premise around how autoprefixer handles grid-areas that each grid has unique area names?

Autoprefixer breaks down immediately if duplicate area names are used as explained in issue https://github.com/postcss/autoprefixer/issues/1038

**Comment by ai** [comment]
Or recommend to duplicate `grid-gap`

**Comment by Dan503** [comment]
I think my proposal is pretty safe. it replicates how the browser will handle the media queries. AUtoprefixer is already restricted to needing each area to be unique (excluding media queries).

**Comment by ai** [comment]
I think it is safe too. Anyway grid logic is not enabled by default. We can expect some limits from grid prefixes users.

**Comment by yepninja** [comment]
@Dan503 Agreed. Mostly you are right. But it would be cool to have feedback from users (how they use autoprefixer and CSS grid).

**Comment by ai** [comment]
They didn't use it because of @rachelandrew article 😣 And since most of CSS stars ignore this feature.

Until @Dan503 will wow a article, we will not have good feedback.

But you can write a tweet and I will retweet it by Autoprefixer account.

**Comment by yepninja** [comment]
@Dan503  Let me help you with article :)
Also, it may be better to write several articles.

**Comment by yepninja** [comment]
I had two speeches at the conferences about CSS Grids with autoprefixer in Russia. So I already have enough contents for an article.

**Comment by Dan503** [comment]
I might be able to get the first article published pretty soon actually. It's pretty much done. I just need to edit it a bit so that it works better with this new 4 part series format that I'm going for now.

**Comment by Dan503** [comment]
@yepninja this is a high priority issue that should get fixed before the article is published.

**Comment by ai** [comment]
Released in 9.1.1
